tools/bridge/end2end-tests/tests/test_transfer.py

```diff
@@ -0,0 +1,427 @@
+
+PARITY_DEV_ACCOUNT = "0x00a329c0648769A73afAc7F9381E08FB43dBEA72"
+PARITY_DEV_KEY = "0x4D5DB4107D237DF6A3D58EE5F70AE63D73D7658D4026F2EEFD2F204C81682CB7"
+
+HOME_RPC_URL = "http://localhost:8545"
+FOREIGN_RPC_URL = "http://localhost:8645"
+
+START_WEI_PER_ACCOUNT = 1 * 10 ** 18
+BRIDGE_FUNDS = 1_000_000 * 10 ** 18
+
+
+class TimeoutException(Exception):
+    pass
+
+
+class Timer:
+    def __init__(self, timeout):
+        self.start_time = None
+        self.timeout = timeout
+
+    def start(self):
+        self.start_time = time.time()
+
+    def is_timed_out(self):
+        if self.start_time is None:
+            raise ValueError("Timer is not started yet")
+        return self.time_passed > self.timeout
+
+    @property
+    def time_passed(self):
+        return time.time() - self.start_time
+
+
+def assert_within_timeout(check_function, timeout, poll_period=0.5):
+    """
+    Runs a check_function with assertions and give it some time to pass
+    :param check_function: The function which will be periodically called. It should contain some assertions
+    :param timeout: After this timeout non passing assertion will be raised
+    :param poll_period: poll interval to check the check_function
+    """
+    timer = Timer(timeout)
+    timer.start()
+
+    while True:
+        try:
+            check_function()
+        except AssertionError as e:
+            if not timer.is_timed_out():
+                time.sleep(poll_period)
+            else:
+                raise TimeoutException(
+                    f"Assertion did not pass after {timeout} seconds. See causing exception for more details."
+                ) from e
+        else:
+            break
+
+
+class Service:
+    def __init__(
+        self,
+        args,
+        *,
+        name=None,
+        env=None,
+        uptest_function=None,
+        timeout=5,
+        poll_interval=0.2,
+        process_settings=None,
+    ):
+        self.args = args
+        self.name = name
+        self.env = env
+        self.timeout = timeout
+        self.poll_interval = poll_interval
+        self._uptest_function = uptest_function
+        self.process = None
+        self._process_settings = process_settings
+
+        if self._process_settings is None:
+            self._process_settings = {}
+
+    def start(self):
+        """Starts the service and wait for it to be up """
+        self.process = Popen(self.args, env=self.env, **self._process_settings)
+        self._wait_for_up()
+        return self.process
+
+    def is_up(self):
+        """Use the uptest to determine if the service is up"""
+        if self._uptest_function is not None:
+            return self._uptest_function
+        else:
+            return True
+
+    def _wait_for_up(self):
+        start_time = time.time()
+        while True:
+            is_up = self.is_up()
+
+            if not is_up:
+                if time.time() - start_time > self.timeout:
```

Comment:
Maybe call `terminate` here as well?

Reply:
yes, you are right, somewhere should be a terminate in cause the program is not up. But it would probably be better in `start`

Reply:
I call terminate in start if the setup failed.